src/mono/mono/mini/simd-intrinsics.c

```diff
@@ -6118,11 +6211,36 @@ mono_simd_decompose_intrinsic (MonoCompile *cfg, MonoBasicBlock *bb, MonoInst *i
 			decompose_vtype_opt_store_arg (cfg, bb, ins, &(ins->dreg));
 	}
 }
+
+gboolean
+mono_simd_unsupported_aggressive_inline_intrinsic_type (MonoMethod *cmethod)
+{
+	/*
+	* If a method has been marked with aggressive inlining, check if we support
+	* aggressive inlining of the intrinsics type, if not, ignore aggressive inlining
+	* since it could end up inlining a large amount of code that most likely will end
+	* up as dead code.
+	*/
```

Comment:
Could you help me understand why this is needed on Mono a bit?

The software fallback for V128/V256/V512 is currently implemented as doing 2x operations on the lower/upper halves (except for a couple of methods like `Shuffle` which operate on the full vector). So `V512` is implemented as `2x V256`, `V256` is implemented as `2x V128`, and `V128` is implemented as `2x V64`. `V64` is then implemented as a loop over the scalar elements with potentially large amounts of generic code.

So I would imagine that on any platform where `V128` is supported, that the codegen for `V256`/`V512` should be generally nice/small, even if aggressively inlined and with no real dead code elimination required. Even in the case where you have an unsupported type like `V256<Guid>`, the first `V128` operation should have a `PNSE` thrown (since whether a type is supported or not is typically a known constant).

So I'd only expect that this is needed for V64 on x86/x64 where there is no acceleration and it has to hit the scalar fallback with the loop over the elements. For RyuJIT this loop isn't an issue due to the generic specialization we do, allowing us to get it down to only the code path that is actually used. I believe this isn't possible for Mono today and is non-trivial to add, so the skipping of inlining does make sense in that regard.

Reply:
So without disabling aggressive inlining I ended up with methods that where huge, and not doing aggressive inline on these types (but still honor inline limit size) made them sane again. I will need to re-iterate around that change in order to tell exactly what happened with our inliner in the aggressive case.

Reply:
Just validated this, with disabling the aggressive inlining as above a .net9 full AOT of S.P.C takes ~1.7 GB of memory, and not doing this will consume an additional 600 MB of memory, so I believe its worth preventing aggressive inlining for these types that are not hardware accelerated on any of the Mono supported platforms. I didn't add V64 since it seems to be handled a little differently on at least ARM case. I won't have bandwidth at the moment to do more deep analysis around why aggressive inlining of these template types cause that large increase in memory so I think doing this change, at least short term is worth it, we could probably file an issue around the bloat of Vector256`1 and Vector512`1, but maybe the better long term solution is to actually implement intrinsic support for these types.

Reply:
For interpreter I'm handling this in a general fashion by detecting early dead pieces of code and not inlining any of the calls there. https://github.com/dotnet/runtime/pull/97514. It is possible that a similar approach for jit can produce further improvements without having to special case classes.

Reply:
Trying to reenable branch optimizations when using llvm in this pr:
https://github.com/dotnet/runtime/pull/97189